src/diff/amfGraphDeltaTypes.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as AmfGraphTypes from "./amfGraphTypes";
+
+/**
+ * External property, added by jsondiffpath to indicate that the node property is an array
+ */
+export const JSON_DIFF_PATCH_ARRAY_KEY = "_t";
+export const JSON_DIFF_PATCH_ARRAY_VALUE = "a";
+
+/**
+ * Type describing the delta of a node property
+ *
+ * Delta is reported in array format
+ * example: [oldValue, newValue, magicalNumber], magicalNumber is used identify removal (0), text diffs(2), movements(3)
+ */
+export type PropertyDelta = unknown[];
+
+/**
+ * Type describing the delta of a node property which is an array
+ */
+export type ArrayDelta = {
+  //key is the index in original array
+  [key: string]: unknown[];
+} & {
+  //"special property to indicate its an array
+  [JSON_DIFF_PATCH_ARRAY_KEY]: string;
+};
+
+/**
+ *  Type describing the delta of a reference node ID
+ */
+export type ReferenceNodeDelta = {
+  /**
+   * Delta is calculated on flattened JSON-LD structure, property referencing other node will just have its id
+   * example delta: {@id:[oldId, newId]}
+   */
+  [AmfGraphTypes.JSON_LD_KEY_ID]: [string, string];
+};
+
+/**
+ * Type describing the delta of a node
+ */
+export type NodeDelta = {
+  [AmfGraphTypes.JSON_LD_KEY_ID]: string;
+  [AmfGraphTypes.JSON_LD_KEY_TYPE]: string[];
+} & {
+  [key: string]: PropertyDelta | ArrayDelta | ReferenceNodeDelta;
+};
+
+/**
+ * Type describing the graph delta reported by jsondiffpatch
+ */
+export type GraphDelta = {
+  /**
+   * key: index in the original amf graph (left graph for removed nodes and right graph for added nodes)
+   * value:  NodeDelta - indicate the properties with in a node are changes
+   *        AmfGraphTypes.Node[]  - indicate a node is added/removed/moved
+   */
+  [key: string]: NodeDelta | AmfGraphTypes.Node[];
```

Comment:
```suggestion
type AddedNodes = AmfGraphTypes.Node[];
type RemovedNodes = AmfGraphTypes.Node[];
export type GraphDelta = {
  [key: string]: NodeDelta | AmfGraphTypes.Node[];
```

Minor change, but I think that this would improve readability when looking at type definitions.

Reply:
not sure.. this feels like added and removed could potentially be different

Reply:
They are different, though, even if it's just conceptually. I'm not too concerned about the name / implementation, though. I just think that the type definition should make it easy to understand that a `GraphDelta` represents changed / added / removed nodes. (Right now it reads to me as "change to a node or a list of nodes", and it's not immediately clear what the list is supposed to be.)

Reply:
type is always same(Node) just the meaning is different. How about type AddedOrRemovedNodes = AmfGraphTypes.Node[];?

Reply:
`AddedOrRemovedNodes` could imply that the things in a single array could be added or removed. Something like `AddedNodesOrRemovedNodes` would be more accurate, but more verbose. (This is why I originally suggested two types, because I couldn't think of a good name for just one.)